### ISSUE TITLE
Improve path definitions

### ISSUE THREAD

**Issue by pydanny** [pydanny/dj-notebook#7]
Would be great if this did at least one of the following:
- Handled paths without user interaction. Not sure if that is possible
- Introspected the manage.py file to determine settings path
- Used some other method to make path management easier, especially for beginners

**Comment by geoffbeier** [comment]
PR #121 shows my thinking on how this might be improved. It looks for settings 3 ways:

1. Explicitly passed to `activate()`. That wins every time.

2. Configured in an environment variable, or in a `.env` file that lives where `load_dotenv()` will find it. I'm not sure this is worthwhile from the perspective of helping beginners, but it's such a common pattern among experienced users that it seems worth trying to support it.

3. It searches for `manage.py` files in the current working directory and in every parent directory until it gets to the root. If it doesn't find one there, it looks 1 level deep into child directories. This should enable a pattern where people create a subdirectory within their project for notebooks or where people keep notebooks at the root of their project. That's the most common way I use them. I feel like maybe we should look more than 1 level deep into child directories, but by default we probably want to cap it... searching too hard through a venv could lead to pain.

If it finds a `manage.py` it uses [`ast`](https://docs.python.org/3/library/ast.html) from stdlib to parse the script and find any calls to `setdefault('DJANGO_SETTINGS_MODULE','name')`. Regex probably would've worked, but this felt better. It then uses [`importlib.util.find_spec`](https://docs.python.org/3/library/importlib.html#importlib.util.find_spec) to see whether the module is loadable and appends the path containing the `manage.py` file where the setting was harvested to `sys.path` if the module is not loadable.

Let's discuss enough to make sure you think the approach is sane and takes this a direction you like. (I think there's enough in place that we should be able to evaluate that.) If so, I'd be happy to clean it up and add more tests to make it mergeable.

**Comment by geoffbeier** [comment]
My only real trick with AST to date has been using it to find anti-patterns in large codebases that I don't know very well. When selective, manual review shows a misuse in one place, and I think it is prolific, I fire up a notebook (which has become my REPL of first resort if I think I'm going to be in there for more than 5 minutes) and exercise the ast module until I think I can identify it reliably, then walk the filesystem trying to do so. This happens to be a very close variant of that trick.

I'll start cleaning this up and add a few more tests. I'll also look around and kind of copy what other code in the package does for docstrings and type annotations. I think adding an optional, keyword-only argument to `activate()` for a dotenv file will make the environment variable easier to use for those who want it, and will make it significantly easier to test.

I was excited to find this project (which I only heard about thanks to the virtual sprint) because it looks like the answer to `shell_plus --notebook` being less brittle and having some nice extra sugar. Getting my django models into a notebook that way had already helped me get things done, and dj-notebook further lowers that friction.

**Comment by geoffbeier** [comment]
I've updated the PR and marked it ready for review, with some comments. I'm pretty happy with its condition at the moment. I'm especially happy with the state of the AST sniffing, even if it's kind of overkill.

I was just testing it more and it now handles even this file properly:

```python
import os


class Environ:
    def setdefault(self, k: str, v: str) -> None:
        print(f"{k}={v}")


environ = Environ()

if __name__ == "__main__":
    environ.setdefault("DJANGO_SETTINGS_MODULE", "foo.bar.settings")
    os.environ.setdefault("DJANGO_SETTINGS_MODULE", "foo.bar.os.environ.settings")

```

Also, that's the first time I've leaned heavily on pycharm's pytest integration. It really made finding/fixing this a snap.

**Comment by geoffbeier** [comment]
I've been using this on a couple of real projects this morning. (The screenshots you've seen have been my toy project for [this series of blog posts](https://geoff.tuxpup.com/posts/hypermedia-systems-django-intro/).)

The workflow for using this as a repl is really, really nice now.

1. From my project's root:
```
poetry add -D dj-notebook
```
(Really, it's a link to my fork right now, but that'll change once this is in a release that hits pypi)

2. Right-click on my project's root in pycharm and add a new notebook called `repl.ipynb`.
3. In the first cell:
```python
from dj_notebook import activate

plus = activate()
```
4. Run the cell.

Now I've got a notebook with all the niceties of pycharm... debugger, etc, all my models hanging off `plus`, and pandas too.

This is the super REPL I wanted `shell_plus --notebook` to be. It is really almost as easy as hitting the REPL tool at the bottom of the pycharm window.

**Comment by pydanny** [comment]
Resolved in #121
